File: C++/hungry_birds.cc
/*
 * hungry_birds.cc - example of concurrent multi-thread application
 *
 * Copyright (C) 2013 Paolo Rovelli
 *
 * Author: Paolo Rovelli <paolorovelli@yahoo.it>
 */

#include <cstdlib>
#include <iostream>
#include <chrono>
#include <thread>
#include <mutex>
#include <condition_variable>

#define NUM_BIRDS 7
#define NUM_WORMS 13

class HungryBirds {
public:
	HungryBirds();
	friend class HungryBirdsBaby;
	friend class HungryBirdsParent;
private:
	int birds;
	int worms;
	std::mutex empty_lock, any_lock;
	std::condition_variable empty, any;
};

class HungryBirdsParent {
public:
	HungryBirdsParent(HungryBirds *hungry);
	void run();
	void join();
private:
	HungryBirds *hungry;
	std::thread thread;
};

class HungryBirdsBaby {
public:
	HungryBirdsBaby(HungryBirds *hungry, int bird);
	void run();
	void join();
private:
	HungryBirds *hungry;
	std::thread thread;
	int bird;
};

/*
 * Hungry Birds - environment
 */
HungryBirds::HungryBirds() : birds(NUM_BIRDS), worms(NUM_WORMS) {
	HungryBirdsParent *parent;
	HungryBirdsBaby *baby[NUM_BIRDS];

	std::cout << std::endl << " + initial dish: " << this->worms
		<< " worms" << std::endl << std::endl;

	parent = new HungryBirdsParent(this);
	for (int i = 0; i < NUM_BIRDS; i++)
		baby[i] = new HungryBirdsBaby(this, i);

	parent->join();
	for (int i = 0; i < NUM_BIRDS; i++)
		baby[i]->join();
}

/*
 * Hungry Birds - parent bird (producer)
 */
HungryBirdsParent::HungryBirdsParent(HungryBirds *hungry) {
	this->hungry = hungry;
	this->thread = std::thread(&HungryBirdsParent::run, this);
}

void HungryBirdsParent::join() {
	this->thread.join();
}

void HungryBirdsParent::run() {
	while(1) {
		std::unique_lock<std::mutex> lock(hungry->empty_lock);
		if (hungry->worms)
			hungry->empty.wait(lock);

		hungry->worms = NUM_WORMS;
		std::cout << " + parent bird brings " << hungry->worms
			<< " worms" << std::endl << std::endl;
		for (int i = 0; i < NUM_WORMS; i++)
			hungry->any.notify_one();
	}
}

/*
 * Hungry Birds - baby birds (consumers)
 */
HungryBirdsBaby::HungryBirdsBaby(HungryBirds *hungry, int bird) {
	this->bird = bird;
	this->hungry = hungry;
	this->thread = std::thread(&HungryBirdsBaby::run, this);
}

void HungryBirdsBaby::join() {
	this->thread.join();
}

void HungryBirdsBaby::run() {
	while(1) {
		std::this_thread::sleep_for(std::chrono::microseconds(
			(rand() % ((1000000*NUM_BIRDS)/NUM_WORMS + 1))));

		std::unique_lock<std::mutex> lock(hungry->any_lock);
		if (!hungry->worms)
			hungry->any.wait(lock);

		hungry->worms--;
		if (hungry->worms) {
			std::cout << " - baby bird " << bird
				<< " eats (dish: " << hungry->worms
				<< " worms)" << std::endl;
		} else {
			std::cout << " - baby bird " << bird
				<< " eats (dish: " << hungry->worms
				<< " worms) and screams"
				<< std::endl << std::endl;
			hungry->empty.notify_one();
		}
	}
}

/*
 * Hungry Birds - main program
 */
int main(int argc, char *argv[])
{
	new HungryBirds();
	return 0;
}

File: Python/hungry_birds.py
#!/usr/bin/env python

"""hungry_birds.py - example of concurrent multi-thread application"""

__copyright__ = "Copyright (C) 2015 Andrea Sindoni, Paolo Rovelli"
__author__ = "Andrea Sindoni, Paolo Rovelli"
__email__ = "paolorovelli@yahoo.it"

import threading
import time
import random

NUM_BIRDS = 7
NUM_WORMS = 13

def parent_bird():
    global worms
    while True:
        empty.acquire()
        with lock:
            try:
                worms = NUM_WORMS
                print " + parent bird brings", worms, "worms\n"
                for i in range(NUM_WORMS):
                    any.release()
            except Exception as e:
                    print e.message

def baby_bird(id):
    global worms
    while True:
        time.sleep(random.random() * NUM_BIRDS/NUM_WORMS)
        any.acquire()
        with lock:
            worms = worms - 1
            try:
                if worms:
                    print " - baby bird", id, "eats (dish:", worms, "worms)"
                else:
                    print " - baby bird", id, "eats (dish:", worms, "worms)" \
                        " and screams\n"
                    empty.release()
            except Exception as e:
                    print e.message

if __name__ == "__main__":

    birds = []
    worms = NUM_WORMS
    print "\n + intitial dish:", worms, "worms\n"

    lock = threading.Lock()
    empty = threading.Semaphore(0)
    any = threading.Semaphore(NUM_WORMS)

    t = threading.Thread(target=parent_bird)
    t.start()
    birds.append(t)

    for i in range(NUM_BIRDS):
        t = threading.Thread(target=baby_bird, args=(i,))
        t.start()
        birds.append(t)

    for t in birds:
        t.join()

